flocker/common/functional/test_script.py

```diff
@@ -272,3 +286,62 @@ def verify_logfiles(stdout_messages, logfile):
         d.addCallback(verify_logfiles, logfile=logfile)
 
         return d
+
+    @if_root
+    def run_journald_script(self, script):
+        """
+        Run a script that logs messages to journald and uses
+        ``FlockerScriptRunner``.
+
+        :param ICommandLineScript: Script to run. Must be class in this module.
+
+        :return: ``Deferred`` that fires with decoded JSON messages from the
+            journal for the process.
+        """
+        name = random_name(self).encode("utf-8")
+        code = _SCRIPT_CODE.format(script.__name__, script.__name__)
+        d = getProcessOutput(
+            b"systemd-run", [b"--unit=" + name, b"--description=testing",
+                             sys.executable, b"-u", b"-c", code,
+                             b"--journald"],
+            env=os.environ, errortoo=True,
+        )
+        d.addCallback(lambda result: msg("systemd-run output: " + result))
+        # Takes a little bit of time before journalctl returns logged messages:
+        d.addCallback(lambda _: time.sleep(3))
+        d.addCallback(lambda _: getProcessOutput(
+            b"journalctl", [b"-u", name, b"-o", b"cat"]))
+        d.addCallback(lambda data: (msg(data), data)[1])
+        d.addCallback(lambda data: [
+            loads(l) for l in data.splitlines() if l.startswith(b"{")])
+        return d
+
+    def test_journald(self):
+        """
+        When ``--journald`` option is used messages end up in the journal.
+        """
+        d = self.run_journald_script(EliotScript)
+        d.addCallback(lambda messages: assertContainsFields(self, messages[1],
+                                                            {u"key": 123}))
+        return d
+
+    def test_journald_percent(self):
+        """
+        When ``--journald`` option is used messages with a ``%`` in them can
+        be logged correctly.
+        """
+        d = self.run_journald_script(EliotPercentScript)
+        d.addCallback(lambda messages: assertContainsFields(
+            self, messages[1], {u"key": u"hello%s"}))
+        return d
+
+    def test_journald_large_message(self):
+        """
+        When ``--journald`` option is used large messages end up in the
+        journal and are not broken up, demonstrating we're not hitting
+        https://bugs.freedesktop.org/show_bug.cgi?id=86465
+        """
+        d = self.run_journald_script(EliotLargeScript)
+        d.addCallback(lambda messages: assertContainsFields(
+            self, messages[1], {u"key": EliotLargeScript.key}))
+        return d
```

Comment:
Ideally we'd have separate tests, instead of just end-to-end tests, yeah... but we do need to test `--journald` option. Given final location of code is in Eliot, which will test the low-level bits on its own, my thought was we should just do the minimal amount of needed testing and make it end-to-end so it's realistic.